tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -486,69 +660,191 @@ void FusionPattern::calculateOperandsAndResults() {
   for (Value v : input_set) operands_.push_back(v);
 }
 
-// Supports using EquivalenceClasses for Value
-bool operator<(const ValueWrapper& lhs, const ValueWrapper& rhs) {
-  auto lhs_value = lhs.getValue().getAsOpaquePointer();
-  auto rhs_value = rhs.getValue().getAsOpaquePointer();
-  return lhs_value < rhs_value;
+// Create a new fusion pattern from a single op.
+FusionPattern::FusionPattern(Operation* op) : FusionPatternBase(op) {
+  if (isRank2RowReduction(op)) {
+    fusion_type_ = FusionType::kRowReduction;
+  } else if (isRank2ColReduction(op)) {
+    fusion_type_ = FusionType::kColReduction;
+  } else if (mlir::disc_ral::isFusible(op)) {
+    fusion_type_ = FusionType::kLoop;
+  } else {
+    fusion_type_ = FusionType::kNone;
+  }
+  dominant_op_ = op;
 }
 
-// shape equality propagation based on the shape constrains of
-// elementwise ops.
-void ShapeConstraintAnalysis::PropagateEquality(
-    const SmallVectorImpl<Operation*>& op_list) {
-  bool converged = true;
-  do {
-    converged = true;
-    auto update = [&](Value lhs, Value rhs,
-                      EquivalenceClasses<ValueWrapper>& impl) {
-      if (!impl.isEquivalent(ValueWrapper(lhs), ValueWrapper(rhs))) {
-        converged = false;
-        impl.unionSets(ValueWrapper(lhs), ValueWrapper(rhs));
-      }
-    };
-    for (Operation* op : op_list) {
-      int num_operand = op->getNumOperands();
-      // Propagates same num_elements equality, and shape equality
-      if (isElementWise(op)) {
-        Value lhs = op->getOperand(0);
-        for (Value rhs : op->getOperands().drop_front()) {
-          update(lhs, rhs, same_num_elements_impl_);
-          update(lhs, rhs, same_shape_impl_);
-        }
-      }
-      // Propagates same num_elements equality, not shape equality
-      if (isa<lmhlo::DynamicReshapeOp, lmhlo::ReshapeOp, lmhlo::TransposeOp>(
-              op)) {
-        Value input = op->getOperand(0);
-        // The last operand is the output memref by design
-        Value output = op->getOperand(num_operand - 1);
-        update(input, output, same_num_elements_impl_);
-      }
-      if (isa<lmhlo::ClampOp>(op)) {
-        Value min = op->getOperand(0);
-        Value operand = op->getOperand(1);
-        Value max = op->getOperand(2);
-        Value output = op->getOperand(3);
-        auto min_ty = min.getType().dyn_cast<MemRefType>();
-        auto operand_ty = operand.getType().dyn_cast<MemRefType>();
-        auto max_ty = max.getType().dyn_cast<MemRefType>();
-        if (!min_ty || !max_ty || !operand_ty) {
+// Create a new fusion pattern from the ops inside the lmhlo fusion op.
+FusionPattern::FusionPattern(lmhlo::FusionOp op, ShapeAnalysis* shape_analysis)
+    : FusionPatternBase(op) {
+  if (shape_analysis != nullptr) {
```

Comment:
What if shape_analysis is nullptr? What's the meaning of such a FusionPattern object?

Reply:
To check.

Reply:
It is correct in current implementation.